File: ZipArchiver.java
// ZipArchiver.java
// ZIP Archiver на Java

import java.io.*;
import java.nio.file.*;
import java.util.*;
import java.util.zip.*;
import java.util.stream.*;

public class ZipArchiver {
    private final String archivePath;

    public ZipArchiver(String archivePath) {
        this.archivePath = archivePath;
    }

    public void create(List<String> files, int level) throws IOException {
        try (FileOutputStream fos = new FileOutputStream(archivePath);
             ZipOutputStream zos = new ZipOutputStream(fos)) {
            zos.setLevel(level);
            for (String f : files) {
                addToZip(zos, Paths.get(f), "");
            }
        }
        System.out.println("Архив " + archivePath + " создан.");
    }

    private void addToZip(ZipOutputStream zos, Path path, String base) throws IOException {
        if (Files.isDirectory(path)) {
            try (Stream<Path> paths = Files.list(path)) {
                for (Path p : paths.collect(Collectors.toList())) {
                    String rel = base.isEmpty() ? p.getFileName().toString() : base + "/" + p.getFileName().toString();
                    addToZip(zos, p, rel);
                }
            }
        } else {
            ZipEntry entry = new ZipEntry(base.isEmpty() ? path.getFileName().toString() : base);
            zos.putNextEntry(entry);
            Files.copy(path, zos);
            zos.closeEntry();
        }
    }

    public void extract(String outputDir) throws IOException {
        try (ZipInputStream zis = new ZipInputStream(new FileInputStream(archivePath))) {
            ZipEntry entry;
            while ((entry = zis.getNextEntry()) != null) {
                Path outPath = Paths.get(outputDir, entry.getName());
                if (entry.isDirectory()) {
                    Files.createDirectories(outPath);
                } else {
                    Files.createDirectories(outPath.getParent());
                    Files.copy(zis, outPath, StandardCopyOption.REPLACE_EXISTING);
                }
                zis.closeEntry();
            }
        }
        System.out.println("Архив " + archivePath + " распакован в " + outputDir);
    }

    public void list() throws IOException {
        try (ZipFile zipFile = new ZipFile(archivePath)) {
            Enumeration<? extends ZipEntry> entries = zipFile.entries();
            while (entries.hasMoreElements()) {
                ZipEntry entry = entries.nextElement();
                System.out.printf("%s  %d байт  %s%n", entry.getName(), entry.getSize(), entry.getTime());
            }
        }
    }

    public void add(List<String> files) throws IOException {
        // Пересоздаём архив с добавлением
        Path temp = Paths.get(archivePath + ".tmp");
        try (ZipOutputStream zos = new ZipOutputStream(Files.newOutputStream(temp))) {
            // Копируем существующие записи
            try (ZipFile zf = new ZipFile(archivePath)) {
                Enumeration<? extends ZipEntry> entries = zf.entries();
                while (entries.hasMoreElements()) {
                    ZipEntry entry = entries.nextElement();
                    zos.putNextEntry(new ZipEntry(entry.getName()));
                    try (InputStream is = zf.getInputStream(entry)) {
                        is.transferTo(zos);
                    }
                    zos.closeEntry();
                }
            }
            // Добавляем новые файлы
            for (String f : files) {
                addToZip(zos, Paths.get(f), "");
            }
        }
        Files.move(temp, Paths.get(archivePath), StandardCopyOption.REPLACE_EXISTING);
        System.out.println("Файлы добавлены в " + archivePath);
    }

    public void remove(List<String> files) throws IOException {
        Path temp = Paths.get(archivePath + ".tmp");
        try (ZipOutputStream zos = new ZipOutputStream(Files.newOutputStream(temp))) {
            try (ZipFile zf = new ZipFile(archivePath)) {
                Enumeration<? extends ZipEntry> entries = zf.entries();
                while (entries.hasMoreElements()) {
                    ZipEntry entry = entries.nextElement();
                    if (files.contains(entry.getName())) {
                        continue;
                    }
                    zos.putNextEntry(new ZipEntry(entry.getName()));
                    try (InputStream is = zf.getInputStream(entry)) {
                        is.transferTo(zos);
                    }
                    zos.closeEntry();
                }
            }
        }
        Files.move(temp, Paths.get(archivePath), StandardCopyOption.REPLACE_EXISTING);
        System.out.println("Файлы удалены из " + archivePath);
    }

    public void comment(String text) throws IOException {
        // Комментарий в Java ZIP не поддерживается (можно добавить как файл .comment)
        System.out.println("Комментарии не поддерживаются в стандартной библиотеке Java.");
    }

    public static void main(String[] args) throws IOException {
        if (args.length == 0 || args[0].equals("help")) {
            System.out.println("Использование: java ZipArchiver <команда> [опции]\n" +
                    "  create -a <archive> -f <files...> [-l level]\n" +
                    "  extract -a <archive> [-o output]\n" +
                    "  list -a <archive>\n" +
                    "  add -a <archive> -f <files...>\n" +
                    "  remove -a <archive> -f <files...>\n" +
                    "  comment -a <archive> -t <text>");
            System.exit(0);
        }

        String command = args[0];
        String archive = null;
        List<String> files = new ArrayList<>();
        String output = ".";
        int level = 6;
        String text = null;

        for (int i = 1; i < args.length; i++) {
            switch (args[i]) {
                case "-a":
                    archive = args[++i];
                    break;
                case "-f":
                    String[] parts = args[++i].split(",");
                    files.addAll(Arrays.asList(parts));
                    break;
                case "-o":
                    output = args[++i];
                    break;
                case "-l":
                    level = Integer.parseInt(args[++i]);
                    break;
                case "-t":
                    text = args[++i];
                    break;
            }
        }

        if (archive == null) {
            System.err.println("Ошибка: укажите -a");
            System.exit(1);
        }

        ZipArchiver archiver = new ZipArchiver(archive);

        switch (command) {
            case "create":
                if (files.isEmpty()) {
                    System.err.println("Ошибка: укажите -f");
                    System.exit(1);
                }
                archiver.create(files, level);
                break;
            case "extract":
                archiver.extract(output);
                break;
            case "list":
                archiver.list();
                break;
            case "add":
                if (files.isEmpty()) {
                    System.err.println("Ошибка: укажите -f");
                    System.exit(1);
                }
                archiver.add(files);
                break;
            case "remove":
                if (files.isEmpty()) {
                    System.err.println("Ошибка: укажите -f");
                    System.exit(1);
                }
                archiver.remove(files);
                break;
            case "comment":
                if (text == null) {
                    System.err.println("Ошибка: укажите -t");
                    System.exit(1);
                }
                archiver.comment(text);
                break;
            default:
                System.err.println("Неизвестная команда");
                System.exit(1);
        }
    }
}

File: ZipArchiver.cs
// ZipArchiver.cs
// ZIP Archiver на C#

using System;
using System.Collections.Generic;
using System.IO;
using System.IO.Compression;
using System.Linq;

class ZipArchiver
{
    private readonly string archivePath;

    public ZipArchiver(string archivePath)
    {
        this.archivePath = archivePath;
    }

    public void Create(IEnumerable<string> files, int level)
    {
        using (var fs = new FileStream(archivePath, FileMode.Create))
        using (var zip = new ZipArchive(fs, ZipArchiveMode.Create))
        {
            foreach (var f in files)
            {
                AddToZip(zip, f, "");
            }
        }
        Console.WriteLine($"Архив {archivePath} создан.");
    }

    private void AddToZip(ZipArchive zip, string path, string basePath)
    {
        if (Directory.Exists(path))
        {
            foreach (var entry in Directory.GetFileSystemEntries(path))
            {
                string rel = basePath == "" ? Path.GetFileName(entry) : Path.Combine(basePath, Path.GetFileName(entry));
                AddToZip(zip, entry, rel);
            }
        }
        else
        {
            string entryName = basePath == "" ? Path.GetFileName(path) : basePath;
            zip.CreateEntryFromFile(path, entryName, CompressionLevel.Optimal);
        }
    }

    public void Extract(string outputDir)
    {
        using (var zip = ZipFile.OpenRead(archivePath))
        {
            foreach (var entry in zip.Entries)
            {
                string path = Path.Combine(outputDir, entry.FullName);
                if (entry.Name == "")
                {
                    Directory.CreateDirectory(path);
                }
                else
                {
                    Directory.CreateDirectory(Path.GetDirectoryName(path));
                    entry.ExtractToFile(path, true);
                }
            }
        }
        Console.WriteLine($"Архив {archivePath} распакован в {outputDir}");
    }

    public void List()
    {
        using (var zip = ZipFile.OpenRead(archivePath))
        {
            foreach (var entry in zip.Entries)
            {
                Console.WriteLine($"{entry.FullName}  {entry.Length} байт  {entry.LastWriteTime}");
            }
        }
    }

    public void Add(IEnumerable<string> files)
    {
        string tempPath = archivePath + ".tmp";
        // Копируем существующие записи
        using (var original = ZipFile.OpenRead(archivePath))
        using (var temp = new FileStream(tempPath, FileMode.Create))
        using (var zip = new ZipArchive(temp, ZipArchiveMode.Create))
        {
            foreach (var entry in original.Entries)
            {
                var newEntry = zip.CreateEntry(entry.FullName, entry.CompressionLevel);
                using (var src = entry.Open())
                using (var dst = newEntry.Open())
                {
                    src.CopyTo(dst);
                }
            }
            // Добавляем новые
            foreach (var f in files)
            {
                AddToZip(zip, f, "");
            }
        }
        File.Move(tempPath, archivePath, true);
        Console.WriteLine($"Файлы добавлены в {archivePath}");
    }

    public void Remove(IEnumerable<string> files)
    {
        string tempPath = archivePath + ".tmp";
        using (var original = ZipFile.OpenRead(archivePath))
        using (var temp = new FileStream(tempPath, FileMode.Create))
        using (var zip = new ZipArchive(temp, ZipArchiveMode.Create))
        {
            foreach (var entry in original.Entries)
            {
                if (files.Contains(entry.FullName))
                    continue;
                var newEntry = zip.CreateEntry(entry.FullName, entry.CompressionLevel);
                using (var src = entry.Open())
                using (var dst = newEntry.Open())
                {
                    src.CopyTo(dst);
                }
            }
        }
        File.Move(tempPath, archivePath, true);
        Console.WriteLine($"Файлы удалены из {archivePath}");
    }

    public void Comment(string text)
    {
        // В .NET нет встроенной поддержки комментариев, но можно добавить файл .comment
        Console.WriteLine("Комментарии не поддерживаются в стандартной библиотеке .NET.");
    }

    static void Main(string[] args)
    {
        if (args.Length == 0 || args[0] == "help")
        {
            Console.WriteLine(@"Использование: ZipArchiver <команда> [опции]
  create -a <archive> -f <files...> [-l level]
  extract -a <archive> [-o output]
  list -a <archive>
  add -a <archive> -f <files...>
  remove -a <archive> -f <files...>
  comment -a <archive> -t <text>");
            return;
        }

        string command = args[0];
        string archive = null;
        List<string> files = new List<string>();
        string output = ".";
        int level = 6;
        string text = null;

        for (int i = 1; i < args.Length; i++)
        {
            switch (args[i])
            {
                case "-a": archive = args[++i]; break;
                case "-f": files.AddRange(args[++i].Split(',')); break;
                case "-o": output = args[++i]; break;
                case "-l": level = int.Parse(args[++i]); break;
                case "-t": text = args[++i]; break;
            }
        }

        if (string.IsNullOrEmpty(archive))
        {
            Console.WriteLine("Ошибка: укажите -a");
            return;
        }

        var archiver = new ZipArchiver(archive);

        switch (command)
        {
            case "create":
                if (!files.Any()) { Console.WriteLine("Ошибка: укажите -f"); return; }
                archiver.Create(files, level);
                break;
            case "extract":
                archiver.Extract(output);
                break;
            case "list":
                archiver.List();
                break;
            case "add":
                if (!files.Any()) { Console.WriteLine("Ошибка: укажите -f"); return; }
                archiver.Add(files);
                break;
            case "remove":
                if (!files.Any()) { Console.WriteLine("Ошибка: укажите -f"); return; }
                archiver.Remove(files);
                break;
            case "comment":
                if (string.IsNullOrEmpty(text)) { Console.WriteLine("Ошибка: укажите -t"); return; }
                archiver.Comment(text);
                break;
            default:
                Console.WriteLine("Неизвестная команда");
                break;
        }
    }
}
